### ISSUE TITLE
Applying function to selected columns with .SD by .I is not working as expected

### ISSUE THREAD

**Issue by teotenn** [Rdatatable/data.table#6013]
# Minimal reproducible example

```
library(data.table)
df <- data.table(A = "First field",
                 B = 1:5,
                 C = NA_character_,
                 D = NA,
                 E = "More fields",
                 Z = "This is not included")
# Cols to include
lookup <- c("A", "B", "C", "D", "E")

df[ , paste(.SD[ , lookup, with=FALSE],collapse=" ! "), by = .I]
```

I am getting the following results:

```
                                                                                                                                                                                                                  V1
1: c("First field", "First field", "First field", "First field", "First field") ! 1:5 ! c(NA, NA, NA, NA, NA) ! c(NA, NA, NA, NA, NA) ! c("More fields", "More fields", "More fields", "More fields", "More fields")
```

**Expected**

```
       I                                      V1
   <int>                                  <char>
1:     1 First field ! 1 ! NA ! NA ! More fields
2:     2 First field ! 2 ! NA ! NA ! More fields
3:     3 First field ! 3 ! NA ! NA ! More fields
4:     4 First field ! 4 ! NA ! NA ! More fields
5:     5 First field ! 5 ! NA ! NA ! More fields
```

I was seeking for help in Reddit and the expected results are from a different user who was trying to help me, running the exact same code above he got what I'm placing in "expected" and I got what I'm showing above. I tried in 2 different computers, with different versions of R and `data.table` and different OS. For my last try I executed directly to an R console (no Rstudio) and I updated to the latest version of `data.table` as you can see below, but I still get the same result.

Thus, I am not sure if this is a bug or I have something wrong. I would really appreciate the help.

# Output of `sessionInfo()`

```
R version 4.3.2 (2023-10-31)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Debian GNU/Linux trixie/sid

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.12.0 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.12.0

locale:
 [1] LC_CTYPE=en_US.UTF-8    LC_NUMERIC=C            LC_TIME=en_US.UTF-8     LC_COLLATE=en_US.UTF-8 
 [5] LC_MONETARY=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 LC_PAPER=en_US.UTF-8    LC_NAME=C              
 [9] LC_ADDRESS=C            LC_TELEPHONE=C          LC_MEASUREMENT=C        LC_IDENTIFICATION=C    

time zone: Europe/Warsaw
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] data.table_1.15.99

loaded via a namespace (and not attached):
[1] compiler_4.3.2 tools_4.3.2    jsonlite_1.8.8
```

**Comment by ben-schwen** [comment]
I get your expected result on R 4.3.2 with latest dev version.

What about if you change your query to `df[ , paste(.SD[ , lookup, with=FALSE],collapse=" ! "), by = seq_len(nrow(df))]` eliminating the possibility that the error is caused by using `by=.I`.

Moreover, what is your output when turning on verbose with
`options(datatable.verbose=TRUE)` before?

It should look like this:
```r
df[ , paste(.SD[ , lookup, with=FALSE],collapse=" ! "), by = seq_len(nrow(df))]
# Finding groups using forderv ... forder.c received 5 rows and 1 columns
# 0.007s elapsed (0.007s cpu)
# Finding group sizes from the positions (can be avoided to save RAM) ... 0.010s elapsed (0.010s cpu)
# lapply optimization is on, j unchanged as 'paste(.SD[, lookup, with = FALSE], collapse = " ! ")'
# GForce is on, but not activated for this query; left j unchanged (see ?GForce)
# Old mean optimization is on, left j unchanged.
# Making each group and running j (GForce FALSE) ...
#   memcpy contiguous groups took 0.000s for 5 groups
#   eval(j) took 0.001s for 5 calls
# 0.000s elapsed (0.001s cpu)
#    seq_len                                      V1
#      <int>                                  <char>
# 1:       1 First field ! 1 ! NA ! NA ! More fields
# 2:       2 First field ! 2 ! NA ! NA ! More fields
# 3:       3 First field ! 3 ! NA ! NA ! More fields
# 4:       4 First field ! 4 ! NA ! NA ! More fields
# 5:       5 First field ! 5 ! NA ! NA ! More fields
```

**Comment by teotenn** [comment]
Thanks, using `seq_len(nrow(df))` works. I don't know why I didn't think about it, I was actually reading that this is the equivalent of `.I`. However, do you have any idea why the `.I` might not be working for me?

When I turn the verbose to TRUE I get almost the same output as you:

```
Finding groups using forderv ... forder.c received 5 rows and 1 columns
0.001s elapsed (0.000s cpu) 
Finding group sizes from the positions (can be avoided to save RAM) ... 0.000s elapsed (0.000s cpu) 
lapply optimization is on, j unchanged as 'paste(.SD[, lookup, with = FALSE], collapse = " ! ")'
GForce is on, left j unchanged
Old mean optimization is on, left j unchanged.
Making each group and running j (GForce FALSE) ... 
  memcpy contiguous groups took 0.000s for 5 groups
  eval(j) took 0.001s for 5 calls
0.001s elapsed (0.001s cpu) 
   seq_len                                      V1
1:       1 First field ! 1 ! NA ! NA ! More fields
2:       2 First field ! 2 ! NA ! NA ! More fields
3:       3 First field ! 3 ! NA ! NA ! More fields
4:       4 First field ! 4 ! NA ! NA ! More fields
5:       5 First field ! 5 ! NA ! NA ! More fields
```

**Comment by ben-schwen** [comment]
It is a bit strange since the use of `by=.I` should also be already available on the latest CRAN version.

Could you try starting a new R session and running `data.table::update_dev_pkg()`?

Your output looks indeed a bit strange since we set the default of the options `datatable.print.class` and `datatable.print.keys` now to `TRUE` and your output seems to not have them. But this should also be already on CRAN.

**Comment by teotenn** [comment]
The issue has been fixed after executing `data.table::update_dev_pkg()` and now the `.I` works as expected. Additionally I can see now the type of each column under the column name (before it did not appear). However, my version of `data.table` is the same (1.15.99) because previously I have installed from github.

So, basically my first try was with a previous CRAN version I had, then I updated from CRAN and finally I removed the installed package and installed again but now from github. All that was before posting here. The result was always the same until now that I ran `data.table::update_dev_pkg()`.
